### PR TITLE
Add `detail::hash_join`

### PR DESCRIPTION
**Pull request by PointKernel**
Closes https://github.com/rapidsai/cudf/issues/10587

This PR adds a `detail::hash_join` class which is templated on the hash function. It also cleans up `join` internal functions by moving code around to proper files. The implementation of `detail::hash_join` is mainly taken from `cudf::hash_join::hash_join_impl`.

### CODE REVIEW

File: cpp/include/cudf/detail/join.hpp
Comment (davidwendt) [inline cpp/include/cudf/detail/join.hpp:151]: There appears to be a special case for this
https://github.com/rapidsai/cudf/pull/10695/files#diff-7e4a3b3c2c7b9cf3ae4c5ef82cd665e8b3579e0599dd3536c754424b000a6e0eR385-R388

Also, does `null == empty` here?  Generally they mean different things. There is an `_is_empty` variable that appears to match with `is null` statements in the documentation.

Comment (davidwendt) [inline cpp/include/cudf/detail/join.hpp:151]: There appears to be a special case for this
https://github.com/rapidsai/cudf/pull/10695/files#diff-7e4a3b3c2c7b9cf3ae4c5ef82cd665e8b3579e0599dd3536c754424b000a6e0eR385-R388

Also, does `null == empty` here?  Generally they mean different things. There is an `_is_empty` variable that appears to match with `is null` statements in the documentation.

**Review by davidwendt — CHANGES_REQUESTED** [review]
Please wait for 2 cpp approvals before merging this PR.

File: cpp/include/cudf/detail/join.hpp
Comment (jrhemstad) [inline cpp/include/cudf/detail/join.hpp:56]: It feels like there's a lot of redundancy between `detail::hash_join` and `hash_join_impl`. 

Can we just merge them and make the PIMPL of the `cudf::hash_join` be a `unique_ptr<detail::hash_join<default_hash>>` ?

Comment (PointKernel) [reply cpp/include/cudf/detail/join.hpp:56]: That's what I tried in the first place. The issue is that I don't want to expose `detail::default_hash` in `cudf/join.hpp` since it's declared in `cudf/detail/utilities/hash_functions.cuh`. Including `hash_functions.cuh` in `cudf/join.hpp` will make the join public header a `.cuh` file as well.

Using `hash_join_impl` as a wrapper to hide `detail::default_hash` can minimize header exposure but like you pointed out, introduces redundancy. Please advise if there is a better solution.

File: cpp/include/cudf/detail/join.hpp
Comment (PointKernel) [reply cpp/include/cudf/detail/join.hpp]: Good catch. I've changed the expression to `is empty`.

File: cpp/include/cudf/detail/join.hpp
Comment (PointKernel) [reply cpp/include/cudf/detail/join.hpp:54]: Fixed by forward-declaration of `MurmurHash3_32`

**Review by jrhemstad — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Pull request merged**
